### PR TITLE
ENH: add two-sample test (Epps-Singleton) to scipy.stats

### PR DESCRIPTION
**Pull request by chrisb83**
Another two-sample test for equality of distributions which can be applied to discrete and continuous distributions and which has high power. 

Output Format: TBD, currently tuple, probably a good idea to change to Bunches, need to find the related discussion.

### CODE REVIEW

**Comment by ev-br** [comment]
Re: related discussion https://github.com/scipy/scipy/pull/6335 and references therein.

File: scipy/stats/_hypotests.py [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+        raise ValueError('x should have at least 5 elements, but len(x) '
+                         'equals {}.'.format(nx))
+    ny = len(y)
+    if ny < 5:
+        raise ValueError('y should have at least 5 elements, but len(y) '
+                         'equals {}.'.format(ny))
+    n = nx + ny
+
+    # rescale t with semi-iqr as proposed in [1]
+    sigma = iqr(np.hstack((x, y))) / 2
```
Comment (rlucas7) [inline scipy/stats/_hypotests.py]: if either or `x` or `y` here has a `None` value the iqr() will throw a runtime warning. Does it make sense to check in this function and throw an error here instead?

File: scipy/stats/_hypotests.py [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+
+    This is a test for the null hypothesis that two samples have the same
+    underlying probability distribution which is based on the empirical
+    characteristic function.
+
+    Parameters
+    ----------
+    x, y : array-like
+        The two samples of observations to be tested. Input must not have more
+        than one dimension. Samples can have different lengths.
```
Comment (rlucas7) [inline scipy/stats/_hypotests.py]: >Input must not have more than one dimension.

This method is supposed to handle discrete cases just like continuous something that KS test doesn't do (according to [2]). I can see a use case with lots of data where forcing the x,y to be array like forces you to create a lot more data. For example, x,y discrete, and the numbers 0,1,2,3,4 occur often so you include weights for each so you'd want x, (e.g.) to be [[x_1, x_2],[w_1, w_2]] where x_1 and x_2 are entries from 0,1,2,3,4 and w_1, w_2 are the occurrences/counts.

File: scipy/stats/_hypotests.py [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+    of the test in general. Table III in [1]_ gives the optimal values for
+    the distributions tested in that study.
+
+    References
+    ----------
+    .. [1] T. W. Epps and K. J. Singleton, "An omnibus test for the two-sample
+       problem using the empirical characteristic function", Journal of
+       Statistical Computation and Simulation 26, p. 177–-203, 1986.
+
+    .. [2] S. J. Goerg and J. Kaiser, "Nonparametric testing of distributions -
```
[the remark this answers is not in the document]
Comment (chrisb83) [reply scipy/stats/_hypotests.py]: [2] is freely accessible, i mention the article in the tests as well

File: scipy/stats/_hypotests.py [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+    This is a test for the null hypothesis that two samples have the same
+    underlying probability distribution which is based on the empirical
+    characteristic function.
+
+    Parameters
+    ----------
+    x, y : array-like
+        The two samples of observations to be tested. Input must not have more
+        than one dimension. Samples can have different lengths.
+    t : array-like, optional
```
[the remark this answers is not in the document]
Comment (chrisb83) [reply scipy/stats/_hypotests.py]: i think in both references they suggest the semi-iqr as a robust scaling

[the remark this answers is not in the document]
Comment (chrisb83) [reply scipy/stats/_hypotests.py]: but agree, let's add it to the documentation

File: scipy/stats/_hypotests.py [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+    cov_y = np.cov(gy.T, bias=True)
+    est_cov = (n/nx)*cov_x + (n/ny)*cov_y
+    est_cov_inv = np.linalg.pinv(est_cov)
+    r = np.linalg.matrix_rank(est_cov_inv)
+    if r < 2*len(t):
+        warnings.warn('Estimated covariance matrix does not have full rank. '
+                      'This indicates a bad choice of the input t and the '
+                      'test might not be consistent.')  # see p. 183 in [1]_
+
+    # compute test statistic w distributed asympt. as chisquare with df=r
```
[the remark this answers is not in the document]
Comment (chrisb83) [reply scipy/stats/_hypotests.py]: yes, [1] has more details on this. we don't have two-sample versions of cramer-von-mises and AD in scipy I think. in any case, detecting such cases is probably out of scope and the user has to decide which test to apply

**Comment by chrisb83** [comment]
Further analysis and tests are in a gist: https://gist.github.com/chrisb83/3a057aebf2cd1d61d81ae0b6c198a7f1

File: scipy/stats/_hypotests.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (rlucas7) [reply scipy/stats/_hypotests.py]: >in any case, detecting such cases is probably out of scope and the user has to decide which 
> test to apply
@chrisb83 fair enough, should this be called out in the docstring (with ref to [1] for details)? 
I could go either way, anytime one uses some code it is always the responsible thing to check whether the code being re-used in appropriate for the task at hand or not.

File: scipy/stats/_hypotests.py (hunk 2 above)
Comment (rlucas7) [reply scipy/stats/_hypotests.py]: @chrisb83 maybe add a #TODO for the weights enhancement somewhere in the code, or do you think the structural specifics of the test does not allow for weight handling?

**Comment by chrisb83** [comment]
I decided to return a named tuple as for the other tests for consistency. Using a `Bunch` would not bring a big advantage in my view (see discussion in #3665).

@rlucas7 I tried to address most of your comments regarding the documentation. 
I don't see that the implementation can be changed easily to handle integers in the way you propose, so I leave that question for the future :)

**Comment by rlucas7** [comment]
> I decided to return a named tuple as for the other tests for consistency. Using a `Bunch` would not bring a big advantage in my view (see discussion in #3665).

I agree with named tuple approach-it works and is clear to read.

> @rlucas7 I tried to address most of your comments regarding the documentation.
> I don't see that the implementation can be changed easily to handle integers in the way you propose, so I leave that question for the future :)

I looked at it closer and I agree. Although weights would be nice to handle, adding that is non-trivial. For example, `iqr` would need to be modified to support weights, something non-trivial by itself. This method would also need to be modified to support weights, something else non-trivial.

Let's shelve the weights idea for now and maybe someone will open ENH pull requests to implement support for weights in the future. 

The pull request LGTM.

**Comment by rgommers** [comment]
Pycodestyle isn't happy with `_hypotests.py`, but unclear why. I checked locally and all looks good. A rebase may fix the CI failures.

File: scipy/stats/_hypotests.py [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+    Test the null hypothesis that two samples have the same underlying
+    probability distribution.
+
+    Parameters
+    ----------
+    x, y : array-like
+        The two samples of observations to be tested. Input must not have more
+        than one dimension. Samples can have different lengths.
+    t : array-like, optional
+        The points (t1, ..., tn) where the empirical characterisitc function is
```
Comment (rgommers) [inline scipy/stats/_hypotests.py]: typo in "characterisitc"

File: scipy/stats/_hypotests.py [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+        to be evaluated. It should a positive distinct numbers. The default
```
Comment (rgommers) [inline scipy/stats/_hypotests.py]: "should a" -> "should be"

**Comment by rgommers** [comment]
I don't have access to [1] and can't find a freely available version. Can someone send me a copy by email?

**Comment by tylerjereddy** [comment]
> Can someone send me a copy by email?

Done, I think

**Comment by rgommers** [comment]
awesome, thanks!

I'll remove these comments again, since they're not particularly useful

File: scipy/stats/_hypotests.py (hunk 1 above)
Comment (rgommers) [reply scipy/stats/_hypotests.py]: Tests with `nan` and `inf` would be useful to add. I just checked `kstest`, and it happily accepts `inf` while choking on `nan` (not by design I think).

**Comment by chrisb83** [comment]
> Pycodestyle isn't happy with _hypotests.py, but unclear why. I checked locally and all looks good. A rebase may fix the CI failures.

yes, i can rebase but i think the PR already passed all tests at some point, not sure why it failed last time

File: scipy/stats/_hypotests.py (hunk 1 above)
Comment (chrisb83) [reply scipy/stats/_hypotests.py]: do you mean checks in this function (e.g. raising an error if there are non-finite values)? maybe i should add `nan_policy` as well.

Comment (chrisb83) [reply scipy/stats/_hypotests.py]: At the moment, nans and inf values trigger "LinAlgError: SVD did not converge", which is not helpful for the user

**Comment by chrisb83** [comment]
@rgommers the failure was due to non-ascii characters. They often happen If I copy the title of the reference from a PDF or website, so good that we have a check for it now.

I also added more checks in the input

**Comment by rgommers** [comment]
Last commits look good to me. I resolved all comments that were addressed. There seem to be two comments of @rlucas7 left that may require a small tweak (or just pushing the resolve button, not sure).

File: scipy/stats/_hypotests.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (chrisb83) [reply scipy/stats/_hypotests.py]: i added the sentence "They recommend the use of the ES test for discrete samples…", so I mark this as resolved

File: scipy/stats/_hypotests.py [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+    if nx < 5:
+        raise ValueError('x should have at least 5 elements, but len(x) '
+                         'equals {}.'.format(nx))
+    ny = len(y)
+    if ny < 5:
+        raise ValueError('y should have at least 5 elements, but len(y) '
+                         'equals {}.'.format(ny))
+    n = nx + ny
+
+    # rescale t with semi-iqr as proposed in [1]
```
[the remark this answers is not in the document]
Comment (chrisb83) [reply scipy/stats/_hypotests.py]: I added "The values of `t` are scaled by the semi-interquartile range in the implementation, see [1]_." I did not see a discussion of other scaling approaches, so I stick to that method. I mark the conversation as resolved

File: scipy/stats/_hypotests.py (hunk 2 above)
Comment (chrisb83) [reply scipy/stats/_hypotests.py]: resolved based on comment https://github.com/scipy/scipy/pull/9539#issuecomment-462575451

**Comment by chrisb83** [comment]
@rgommers @rlucas7 I marked all conversations as "resolved" (including a comment why the feedback of rlcuas is addresed).

**Comment by rgommers** [comment]
Okay thanks @chrisb83, in it goes! And thanks @rlucas7 for reviewing

**Comment by rgommers** [comment]
@chrisb83 could you add a release note at https://github.com/scipy/scipy/wiki/Release-note-entries-for-SciPy-1.3.0?

**Pull request merged**